src/Database/PG/Query/Class.hs

```diff
@@ -75,12 +78,22 @@ newtype SingleRow a = SingleRow
 type AltJ :: Type -> Type
 newtype AltJ a = AltJ {getAltJ :: a}
 
-mapEither :: (a -> c) -> (b -> d) -> Either a b -> Either c d
-mapEither f _ (Left a) = Left $ f a
-mapEither _ f (Right b) = Right $ f b
-
+-- | because of CockroachDB we don't know whether we're getting JSON or JSONB
+-- data back. Therefore we try one and then the other.
 instance (FromJSON a) => FromCol (AltJ a) where
-  fromCol = fromColHelper (PD.fn $ mapEither fromString AltJ . eitherDecodeStrict)
+  fromCol input =
+    case decodeJsonb of
+      Right result -> pure result
+      _ -> decodeJson
```

Comment:
Hm. This is somewhat suboptimal I think. Consider how this fails if we get a jsonb payload that fails to parse (because one of PG or Aeson is broken or disagree for whatever reason)?

Consider also the performance of json(non-b) payloads. All `PD.jsonb_ast` appears to do is discard the first byte and then proceed the same as `PD.json_ast`

<details>
<summary>(excerpt from postgresql-binary)</summary>

```
-- |
-- Given a function, which parses a plain UTF-8 JSON string encoded as a byte-array,
-- produces a decoder.
{-# INLINABLE json_bytes #-}
json_bytes :: (ByteString -> Either Text a) -> Value a
json_bytes cont =
  getAllBytes >>= parseJSON
  where
    getAllBytes =
      BinaryParser.remainders
    parseJSON =
      either BinaryParser.failure return . cont

{-# INLINABLE jsonb_ast #-}
jsonb_ast :: Value Aeson.Value
jsonb_ast =
  jsonb_bytes $ mapLeft fromString . Aeson.eitherDecodeStrict'

-- |
-- Given a function, which parses a plain UTF-8 JSON string encoded as a byte-array,
-- produces a decoder.
-- 
-- For those wondering, yes,
-- JSONB is encoded as plain JSON string in the binary format of Postgres.
-- Sad, but true.
{-# INLINABLE jsonb_bytes #-}
jsonb_bytes :: (ByteString -> Either Text a) -> Value a
jsonb_bytes cont =
  getAllBytes >>= trimBytes >>= parseJSON
  where
    getAllBytes =
      BinaryParser.remainders
    trimBytes =
      maybe (BinaryParser.failure "Empty input") return .
      fmap snd . ByteString.uncons
    parseJSON =
      either BinaryParser.failure return . cont
```

</details>

So the worst case we can imagine is decoding a payload `[{.... unbounded object ... }]` which will parse along merryly as jsonb right until the last byte where it fails, and then we do it all again as plain json(non-b).

It's a bit of a shame the protocol is like this. I certainly have to stretch my imagination to find reasons to artificially distinguish encoding formats that are otherwise identical.

It's also sad that this is only directed by haskell types. If there were some metadata header of column transport encodings we could select parsers based on that.

I think the best we can do is probably to just inspect the first byte and discard it if it is `\1` and then continue parsing as JSON.

Reply:
Yeah, that makes sense. Have made the suggested change, all seems to be working.